stdlib/public/Concurrency/ExecutorAssertions.swift

```diff
@@ -11,6 +11,156 @@
 //===----------------------------------------------------------------------===//
 
 import Swift
+import SwiftShims
+
+// ==== -----------------------------------------------------------------------
+// MARK: Precondition executors
+
+/// Unconditionally if the current task is executing on the expected serial executor,
+/// and if not crash the program offering information about the executor mismatch.
+///
+/// This function's effect varies depending on the build flag used:
+///
+/// * In playgrounds and `-Onone` builds (the default for Xcode's Debug
+///   configuration), stops program execution in a debuggable state after
+///   printing `message`.
+///
+/// * In `-O` builds (the default for Xcode's Release configuration), stops
+///   program execution.
+///
+/// * In `-Ounchecked` builds, the optimizer may assume that this function is
+///   never called. Failure to satisfy that assumption is a serious
+///   programming error.
+///
+/// - Parameter executor: the expected current executor
+@available(SwiftStdlib 5.9, *) // FIXME: use @backDeploy(before: SwiftStdlib 5.9)
+public
+func preconditionTaskOnExecutor(
+    _ executor: some SerialExecutor,
+    message: @autoclosure () -> String = String(),
+    file: StaticString = #fileID, line: UInt = #line
+) {
+  let expectationCheck = _taskIsCurrentExecutor(executor.asUnownedSerialExecutor().executor)
+
+  /// TODO: implement the logic in-place perhaps rather than delegating to precondition()?
+  precondition(expectationCheck,
+      // TODO: offer information which executor we actually got
+      "Incorrect actor executor assumption; Expected '\(executor)' executor. \(message())",
+      file: file, line: line) // short-cut so we get the exact same failure reporting semantics
+}
+
+/// Unconditionally if the current task is executing on the serial executor of the passed in `actor`,
+/// and if not crash the program offering information about the executor mismatch.
+///
+/// This function's effect varies depending on the build flag used:
+///
+/// * In playgrounds and `-Onone` builds (the default for Xcode's Debug
+///   configuration), stops program execution in a debuggable state after
+///   printing `message`.
+///
+/// * In `-O` builds (the default for Xcode's Release configuration), stops
+///   program execution.
+///
+/// * In `-Ounchecked` builds, the optimizer may assume that this function is
+///   never called. Failure to satisfy that assumption is a serious
+///   programming error.
+///
+/// - Parameter actor: the actor whose serial executor we expect to be the current executor
+@available(SwiftStdlib 5.9, *) // FIXME: use @backDeploy(before: SwiftStdlib 5.9)
+public
+func preconditionTaskOnActorExecutor(
+    _ actor: some Actor,
+    message: @autoclosure () -> String = String(),
+    file: StaticString = #fileID, line: UInt = #line
+) {
+  guard _isDebugAssertConfiguration() || _isReleaseAssertConfiguration() else {
```

Comment:
I these need to be inlinable or AEIC, otherwise you're picking up the configuration used to build the Concurrency library rather than the configuration used to build the code calling these functions.

Reply:
Ah good catch, thank you

Reply:
Is `@_transparent` a better option? seems that's what assert does 🤔

Reply:
Probably, I don't know those keywords super well.

Reply:
I might follow up on this later then when I had the chance to figure that detail out 👍